File: Day 3 (removeKfromList).hs
-- Definition for singly-linked list:
-- data ListNode a = ListNode { val :: a
--                            , next :: ListNode a
--                            } | Nil deriving Show
--
removeKFromList :: Eq a => ListNode a -> a -> ListNode a
removeKFromList (ListNode val next) k
    |val == k = removeKFromList next k -- if the value is the same, move on without including it
    |otherwise = (ListNode val (removeKFromList next k)) -- otherwise return the same thing but recursively checking the next item
removeKFromList Nil k = Nil

File: Day 3 (removeKfromList).java
// Definition for singly-linked list:
// class ListNode<T> {
//   ListNode(T x) {
//     value = x;
//   }
//   T value;
//   ListNode<T> next;
// }
//

//note: this crashes via stack overflow for large lists
// reimplementation of Haskell algorithm, which doesn't work well in Java
ListNode<Integer> removeKFromList(ListNode<Integer> l, int k) {
    if (l==null)
        return null;
    if (l.value==k&&l.next!=null)
        return removeKFromList(l.next, k);
    if (l.value==k&&l.next==null)
        return null;
    ListNode<Integer> temp = new ListNode<Integer>(l.value);
    if (l.next==null)
        return temp;
    temp.next=removeKFromList(l.next, k);
    return temp;
}
